src/module/wallet/utils/WalletController.ts

```diff
@@ -0,0 +1,245 @@
+        await WalletStorage.setUnencryptedWallets(storageWallets, network);
+
+        return { wallets: newWallets };
+    }
+
+    /**
+     * Recovers the wallets corresponding to the mnemonic or the privateKeys stored in the secure storage
+     * Observations:
+     * - Some new wallets can be created outside the app
+     * - Some wallets can be deleted outside the app
+     * @returns The recovered wallets
+     */
+    static async recoverWallets(network: NetworkType): Promise<WalletsControllerBaseReturn> {
+        const secureStorage = await WalletStorage.getSecure();
+        //Info about the wallets (that is set into the state)
+        const oldStorageWallets = await WalletStorage.getWallets(network); //Sorted by index
+        //Has all the privateKeys and walletIds that point into the oldStorageWallets
+        const oldWalletGroups = secureStorage?.[network] || [];
+        const mainPrivateKey = secureStorage?.mainPrivateKey || ""; //If has previous wallets, it has a mainPrivateKey
+
+        const newWallets: Wallet[] = [];
+        let numberOfNewAccounts = 0;
+        const newWalletGroups: TempWalletGroup[] = [];
+        const deletedIds: number[] = [];
+
+        //Does not have any previous wallet
+        if (oldStorageWallets.length === 0 || oldWalletGroups.length === 0) return { wallets: [] };
+
+        for (const walletGroup of oldWalletGroups) {
+            const tempWallets: Wallet[] = [];
+            const accountDeletedIds: number[] = [];
+
+            //Get all the accounts from the private key
+            const accounts = await ServiceInstance.addServiceInstances({ network: network, privateKey: walletGroup.privateKey });
+            const imported = walletGroup.privateKey !== mainPrivateKey;
+            //Recover the old accounts and check if there are deleted accounts
+            for (const walletId of walletGroup.walletIds) {
+                const wallet = WalletUtils.getWallet(walletId, oldStorageWallets);
+                if (!wallet) {
+                    /* eslint-disable no-console */
+                    console.warn("Corrupted storage: Wallet not found. WalletId: ", walletId);
+                } else if (accounts.find((a) => a.account === wallet.account)) {
+                    const newWallet = {
+                        ...wallet,
+                        colorIndex: WalletUtils.getWalletColor(wallet.account),
+                        ...(imported && { imported: true }),
+                    };
+                    newWallets[wallet.index] = newWallet;
+                    tempWallets.push(newWallet);
+                } else {
+                    deletedIds.push(walletId);
+                    accountDeletedIds.push(walletId);
+                }
+            }
+
+            //Check if there are new accounts
+            const newTempAccounts = accounts.filter(({ account }) => !tempWallets.find((w) => w.account === account));
+            numberOfNewAccounts += newTempAccounts.length;
```

Comment:
Això potser es pot anara augmentant dins del for, així no cal calcular el newTempAccounts que pot arribar a ser bastant costos si tens varies accounts i varios wallets